Ydoc snapshot on server

nonumpa opened the pull request:
### Ydoc snapshot 
  - Snapshot when there's a connection closed (if we snapshot onStoreDocument, it will create too mush snapshots)
  - Instead of using [Y.equalSnapshots](https://github.com/yjs/yjs/blob/5ee6992d1fce68993b9a6ea216358d27fc29b294/src/utils/Snapshot.js#L52) which compares whole snapshot including some unnecessary field such as (connected)`users`, we use custom `equalSnapshots` to compare only prosemirror content.
  - [migrate] `index.afterLoadDocument` snapshot for those article had transcript already but didn't have version snapshots

### Migration
- To add a default snapshot to existing ydocs
  Make sure that rumors-db re-indexed to v1.0.1, then run
  ```
  npx ts-node --esm .\migration\migrate_to_v1.0.1.ts
  ```

### Other changes
- Extract storeArticleText logic from elasticsearch extension to index.ts

nonumpa committed:
Implement ydoc snapshot function which executed when there's connection closed

nonumpa committed:
Extract storeArticleText logic from elasticsearch extension to index.ts

nonumpa committed:
Snapshot for those article had transcript already but didn't have version snapshots

johnson-liang-appier commented on src/index.ts:
```diff
@@ -1,16 +1,88 @@
+    const text = docToPlainText(data.document);
+
+    const db = new elasticsearch.Client(
+      elasticsearchOpts || {
```
Nit: it seems that `elasticsearchOpts` is defined above and always exist, this the setting after `||` is never used.

nonumpa replied:
fixed

MrOrz commented on src/snapshot.ts:
```diff
@@ -0,0 +1,53 @@
+  const versions: Y.Array<{
+    date: number;
+    snapshot: Uint8Array;
+  }> = doc.getArray('versions');
```
Discussion on Slack: https://g0v-slack-archive.g0v.ronny.tw/index/channel/C2PPMRQGP#ts-1691721806.3289

- Does not see other code examples putting snapshot into ydoc that is being snapshotted
- We may consider storing `versions` into Elasticsearch doc
    - We may consider passing the same `elasticsearchOpts` to our snapshot extension to avoid coupling with  
`hocuspocus-extension-elasticsearch` -- the snapshot extension can talk to Elasticsearch by itself

nonumpa replied:
fixed

nonumpa committed:
Refactor ydoc versions, store it in elasticsearch

- binary data stores in base64 string
- rename version.date to version.createdAt

nonumpa committed:
Import elasticsearch package

nonumpa committed:
Use update function instead of index to avoid other fields being overridden

nonumpa committed:
Remove unused default value

MrOrz commented on src/index.ts:
```diff
@@ -1,16 +1,84 @@
+const afterLoadDocument = async ({ document }: afterLoadDocumentPayload) => {
+  // migration script, for those article had transcript already but didn't have version snapshots
+  if (
+    document.getArray('versions').length === 0 &&
+    document.getXmlFragment('prosemirror').toJSON()
```
Question: as `versions` are moved to ES index instead of document, does this condition still work as expected?

nonumpa replied:
Replaced with a migration script.

MrOrz commented on src/snapshot.ts:
```diff
@@ -0,0 +1,120 @@
+    });
+  } catch (e) {
+    if (!e.meta) {
+      console.error(e);
```
Suggest adding prefix so that we know the error console is from snapshot

```suggestion
      console.error('[snapshot]', e);
```

Also for the `console.error` in `else` block

nonumpa replied:
Fixed

nonumpa committed:
Migrate db from v1.0.0 to v1.0.1

nonumpa committed:
FIx log, add prefix so that we can exactly know where the error from

MrOrz commented:
Docker image doesn't seem to start
<img width="954" alt="圖片" src="https://github.com/cofacts/collab-server/assets/108608/1971e1b9-be75-41db-b6c9-cd1bb49fab5b">

nonumpa committed:
Fix cannot find snapshot module issue

ref: https://stackoverflow.com/questions/62619058/appending-js-extension-on-relative-import-statements-during-typescript-compilat

MrOrz commented:
Weirdly enough, I am seeing snapshot.js inside /srv/www/dist/src
![image](https://github.com/cofacts/collab-server/assets/108608/4adc7972-2ea0-413b-bd34-362fdce81b0b)

nonumpa commented:
It seems that the problem is TypeScript is limited to support extension-less `import` statements.
One way is to add .js extension: `./snapshot.js` https://stackoverflow.com/questions/62619058/appending-js-extension-on-relative-import-statements-during-typescript-compilat

johnson-liang-appier reviewed (approved):
[The commit](https://github.com/cofacts/collab-server/pull/4/commits/6baba1111308f9ad2d7efd0836e9a38dcb9d8b5e) works and it's on staging now. Thanks for the update!

MrOrz commented:
Successfully run migration on staging
![image](https://github.com/cofacts/collab-server/assets/108608/a26e2682-1e8e-418c-b7a0-b838f1382fee)
